bitonic.hpp

```diff
@@ -0,0 +1,375 @@
+                {
+                    T tmp = shared[j];
+                    shared[j] = shared[i];
+                    shared[i] = tmp;
+                }
+            }
+        }
+        else
+        {
+            for(unsigned k = 0; k < (thread_elems + 1) / 2; ++k)
+            {
+                const unsigned thread_first_idx = thread_elems / 2 * threadIdx.x;
+                // Here, tier is within the elements the block is responsible for
+                const unsigned tier = (thread_first_idx + k) / (1 << (pass - subpass - 1));
+                const unsigned column = (thread_first_idx + k) % (1 << (pass - subpass - 1));
+                const unsigned i = tier * (1 << (pass - subpass)) + column;
+                const unsigned j = i + (1 << (pass - subpass - 1));
+
+                if((block_first_i + j) < len && shared[j] < shared[i])
+                {
+                    T tmp = shared[j];
+                    shared[j] = shared[i];
+                    shared[i] = tmp;
+                }
+            }
+
+            __syncthreads();
+        }
+    }
+}
+
+template<typename T>
+__global__ void finish_bitonic_step_b_shared(T* arr, unsigned pass, unsigned first_subpass, unsigned len)
+{
+    extern __shared__ T shared[];
+
+    const unsigned block_elems = 1 << (pass - first_subpass);
+    const unsigned block_first_i = block_elems * blockIdx.x;
+    const unsigned thread_elems = (block_elems + blockDim.x - 1) / blockDim.x;
+
+    if((block_first_i + thread_elems * threadIdx.x) < len)
+    {
+        std::memcpy(&shared[thread_elems * threadIdx.x], &arr[block_first_i + thread_elems * threadIdx.x], ::my_min(thread_elems, len - (block_first_i + thread_elems * threadIdx.x)) * sizeof(T));
+    }
+
+    __syncthreads();
+
+    bitonic_step_b_shared(shared, block_elems, pass, first_subpass, len);
+
+    __syncthreads();
+
+    if((block_first_i + thread_elems * threadIdx.x) < len)
+    {
+        std::memcpy(&arr[block_first_i + thread_elems * threadIdx.x], &shared[thread_elems * threadIdx.x], ::my_min(thread_elems, len - (block_first_i + thread_elems * threadIdx.x)) * sizeof(T));
+    }
+}
+
+template<typename T>
+__global__ void bitonic_kernel_shared(T* arr, unsigned passes_to_do, unsigned len)
+{
+    extern __shared__ T shared[];
+
+    const unsigned block_elems = 1 << passes_to_do;
+    const unsigned block_first_i = block_elems * blockIdx.x;
+    const unsigned thread_elems = (block_elems + blockDim.x - 1) / blockDim.x;
+
+    if((block_first_i + thread_elems * threadIdx.x) < len)
+    {
+        std::memcpy(&shared[thread_elems * threadIdx.x], &arr[block_first_i + thread_elems * threadIdx.x], ::my_min(thread_elems, len - (block_first_i + thread_elems * threadIdx.x)) * sizeof(T));
+    }
+
+    __syncthreads();
+
+    for(unsigned pass = 0; pass < passes_to_do; ++pass)
+    {
+        // step a
+        if(thread_elems >= (1 << (pass + 1))) // thread can do entire sub-problem
```

Comment:
I'm pretty sure there's a more elegant and/or less verbose/repetitive way to code this logic for both cases, but I don't think there is a performance implication to doing it this way since all threads in a warp will pick the same branch every time.